content/blog/package-react-components/index.md

```diff
@@ -0,0 +1,819 @@
+---
+title: Package, Publish and Distribute React Components in NPM
+date: "2020-01-05T22:12:03.284Z"
+author: franleplant
+description:
+  "In this post we are going to cover the steps we took to create the nosleepjavascript.com blog
+  in simple, easy, steps using the latest front end technologies: Gatsby.js (JAMstack), NodeJs, React,
+  GraphQL, Javascript, GitHub Pages and GitHub Actions"
+todo:
+  - in the future it would be nice to add a section on to how to deploy to azure
+tags:
+  - gatsbyjs
+  - javascript
+  - react
+---
+
+## TODO
+
+- very basic introduction to webpack (maybe a separate post?)
+
+Keywords
+
+How to publish react components to npm
+How to distribute react components over npm
+How to package react components
+How to shared react components between different apps
+How to package react components and assets such as css and images
+Bundling and sharing react components with Webpack
+
+
+
+
+## What?
+
+There are diverse situation in which we want to create an NPM package that contains
+a [React][1] Component with css and images.
+
+This situation is particularly common inside companies that want to share components,
+and pieces of code in general, between apps usually inside a private NPM repository, but
+it also apply to Open Source software like a single Component package such as a Date Picker
+or a Multi Select and also to Component Libraries or UI Frameworks like [React Bootstrap][9],
+[Ant Design][8], [Semantic UI][4], [Material UI][7], etc (more on these later).
+
+This problem might seem simple enough but it turns out that it has a lot of nuances and trade-offs.
+
+The main challenge in packaging a [React][1] Component is surprisingly not Javascript,
+but all the Assets that might be related to that component such as CSS, Images, SVGs and fonts.
+These assets are directly related to our Javascript code via some of the following statements:
+
+- `imports image from './image.{png,jpg,svg}'` in Javascript
+- `import classNames from './styles.{css,scss,less)'` in Javascript
+- `url(image.{png,jpg,svg})` in CSS
+- `font-family and @import` in CSS
+
+[Webpack][2] can understand all these dependency statements via _loaders_ and completely
+analyse the cross-language and cross-file-type dependency graph, which for a very simple
+component might look something like this.
+
+
+```
+            --> styles.css --> otherImage.svg
+MyComponent --> image.svg
+
+```
+
+It turns out that there are a lot of schemes for representing this graph in a Packaged NPM module,
+each with it's own trade-offs. Analysing each alternative is the main purpose of this post.
+
+Unfortunately in early 2020 there still no universal standard way of Packaging a Front End module with all
+its assets, but a lot of progress has been made and there are actually lots of promising solutions.
+
+We are going to explore how this problem is tackled in the wild,
+we will see case studies and at the end of this post we are going
```

Comment:
TODO: remove case studies (since I completely remove that part) and instead talk about how we related these techniques to real use cases, in each technique we talk about libs that use them in the wild (which is one of the reasons I remove a dedicated case studies section)